link-hint.el

```diff
@@ -1262,12 +1262,16 @@ First try `apply'. If there is an error (ARGS is the wrong number of arguments
 for FUNC), `funcall' FUNC with ARGS. Finally, call FUNC alone."
   (when parser
     (setq args (funcall parser args action)))
+  (unless (listp args)
```

Comment:
Can you remove these lines? I don't think they are necessary/add anything.

Reply:
When I remove these lines, I get the following error with the example snippet from #224:

```
Debugger entered--Lisp error: (wrong-type-argument listp "https://google.com/")
  apply(identity "https://google.com/")
  (condition-case nil (apply func args) (wrong-number-of-arguments (condition-case nil (funcall func args) (wrong-number-of-arguments (funcall func)))))
  link-hint--apply(identity "https://google.com/" link-hint--process-url :describe)
  (let* ((original-win (get-buffer-window)) link-buffer link-buffer-original-pos (link-pos (plist-get link :pos)) (link-win (plist-get link :win)) new-win-buffer (type (plist-get link :type)) (parser (get type :parse)) (args (plist-get link :args)) (link-description (link-hint--apply (or (get type :describe) #'identity) args parser :describe)) ret) (select-window link-win) (progn (setq link-buffer (current-buffer)) (setq link-buffer-original-pos (point))) (goto-char link-pos) (setq ret (link-hint--apply (get type action) args parser action)) (if link-hint-restore (progn (setq new-win-buffer (window-buffer (selected-window))) (cond ((and (eq new-win-buffer link-buffer) (= (point) link-pos)) (goto-char link-buffer-original-pos) (if (window-valid-p original-win) (progn (select-window original-win)))) ((and (buffer-live-p link-buffer) (not (eq new-win-buffer link-buffer))) (if (and (window-valid-p link-win) (eq link-buffer ...)) (set-window-point link-win link-buffer-original-pos) (save-excursion (save-current-buffer ... ...))))))) (link-hint--message action link-description type) ret)
  link-hint--action(:open (:pos 4 :win #<window 951 on *scratch*> :args "https://google.com/" :type link-hint-text-url))
  (progn (link-hint--action action link))
<snip>
```

It seems like, at least in some cases, `link-hint--apply` is passed a non-list value for `args`, which would break `apply` in this way.  

I guess, even better would be to avoid doing this fallback behavior (thus more-or-less deleting `link-hint--apply`) and always require that `func` takes a single argument (and just adapt existing callsites if necessary).

Reply:
Ah, yeah it's not a wrong-number-of-arguments error.  I agree that the current way link-hint is handling this not good, but it's not reliable to just catch wrong-number-of-arguments errors like this.  For example, if there was some function that took an alist or list of lists as a single argument, you could potentially pass it something incorrect but it would fail with a different error.

It looks like all existing :at-point-p functions return a single argument and all :open/actions take 1 or no arguments, so I don't think we actually need the apply part at all.  I'd still like to support 0 argument functions automatically as a fallback though.  I'll have look over the actions, but I think this wouldn't require changing any of the definitions.

I know some Emacs starter kit/distro is using link-hint, so I'm hesitant to break backwards compatibility.  Probably it isn't making custom definitions though, in which case this should affect no one or very few people.  I'll check on that.

Could you try just removing the apply step altogether?  Instead it will try to funcall with 1 argument or fallback to 0 arguments.  Could you also update the "Action Keywords" documentation so that it no longer says you can use multiple arguments (1 or none only now)?